File: backend/apps/energy_gateway/src/fred_feed_server.erl
%% @doc Federal Reserve FRED macro data feed GenServer.
%%
%% Fetches macro indicator series from the FRED API that contextualise
%% energy price movements: USD index, CPI, real interest rates, and
%% industrial production.
%%
%% == Series polled ==
%%
%% <ul>
%%   <li>`DTWEXBGS'  — Nominal Broad USD Index (daily)</li>
%%   <li>`CPIAUCSL'  — CPI All Urban Consumers (monthly)</li>
%%   <li>`REAINTRATREARAT10Y' — 10-yr real interest rate (daily)</li>
%%   <li>`INDPRO'    — Industrial Production Index (monthly)</li>
%%   <li>`DCOILWTICO' — WTI spot (cross-check / baseline)</li>
%% </ul>
%%
%% API key set via `FRED_API_KEY' environment variable.
-module(fred_feed_server).
-behaviour(gen_server).

-export([start_link/0, force_poll/0]).
-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
         terminate/2, code_change/3]).

-define(SERVER,   ?MODULE).
-define(BASE_URL, "https://api.stlouisfed.org/fred/series/observations").
-define(ENDPOINT, fred).

-define(SERIES_IDS, [
    "DTWEXBGS",
    "CPIAUCSL",
    "REAINTRATREARAT10Y",
    "INDPRO",
    "DCOILWTICO"
]).

%%=============================================================================
%% Public API
%%=============================================================================

-spec start_link() -> gen_server:start_ret().
start_link() ->
    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).

-spec force_poll() -> ok.
force_poll() ->
    gen_server:cast(?SERVER, poll).

%%=============================================================================
%% GenServer callbacks
%%=============================================================================

-spec init([]) -> {ok, map()}.
init([]) ->
    ApiKey = os:getenv("FRED_API_KEY", ""),
    %% FRED data updates daily; poll every 4 hours
    erlang:send_after(4000, self(), poll),
    lager:info("FRED feed started (key present: ~p)", [ApiKey =/= ""]),
    {ok, #{interval => 14_400_000, api_key => ApiKey}}.

handle_call(_Req, _From, State) -> {reply, ok, State}.
handle_cast(poll, State) -> do_poll(State), {noreply, State}.

handle_info(poll, #{interval := Interval} = State) ->
    do_poll(State),
    erlang:send_after(Interval, self(), poll),
    {noreply, State};

handle_info(_Info, State) -> {noreply, State}.

terminate(_Reason, _State) -> ok.
code_change(_OldVsn, State, _Extra) -> {ok, State}.

%%=============================================================================
%% Internal helpers
%%=============================================================================

-spec do_poll(State :: map()) -> ok.
do_poll(#{api_key := ""}) ->
    lager:info("FRED API key absent; skipping"),
    ok;
do_poll(#{api_key := Key}) ->
    lists:foreach(fun(S) ->
        case rate_limiter:check_and_consume(?ENDPOINT) of
            ok               -> fetch_series(S, Key);
            {error, _}       -> lager:warning("FRED rate limited for ~s", [S])
        end
    end, ?SERIES_IDS).

-spec fetch_series(SeriesId :: string(), ApiKey :: string()) -> ok.
fetch_series(SeriesId, ApiKey) ->
    Url = ?BASE_URL ++ "?series_id=" ++ SeriesId ++
          "&api_key=" ++ ApiKey ++
          "&file_type=json&sort_order=desc&limit=1",
    case hackney:get(Url, [], <<>>, [with_body]) of
        {ok, 200, _, Body} ->
            parse_and_publish(SeriesId, Body);
        {ok, Status, _, _} ->
            lager:warning("FRED ~p returned HTTP ~p", [SeriesId, Status]);
        {error, Reason} ->
            lager:error("FRED ~p error: ~p", [SeriesId, Reason])
    end.

-spec parse_and_publish(SeriesId :: string(), Body :: binary()) -> ok.
parse_and_publish(SeriesId, Body) ->
    try
        D          = jsx:decode(Body, [return_maps]),
        [Obs | _]  = maps:get(<<"observations">>, D, []),
        Value      = maps:get(<<"value">>,  Obs, <<".">>),
        Date       = maps:get(<<"date">>,   Obs, <<"">>),
        SeriesBin  = list_to_binary(SeriesId),
        %% FRED uses "." for missing; convert to null
        NumValue   = case Value of
            <<".">> -> null;
            V       -> binary_to_float_safe(V)
        end,
        Payload = #{series => SeriesBin, value => NumValue, date => Date},
        ets_cache:put({macro, SeriesBin}, Payload),
        data_router:publish(#{
            type      => macro,
            source    => fred,
            symbol    => SeriesBin,
            timestamp => erlang:system_time(millisecond),
            payload   => Payload
        })
    catch C:E ->
        lager:error("FRED parse error ~p: ~p:~p", [SeriesId, C, E])
    end.

-spec binary_to_float_safe(B :: binary()) -> float() | null.
binary_to_float_safe(B) ->
    try
        case binary:match(B, <<".">>) of
            nomatch -> float(binary_to_integer(B));
            _       -> binary_to_float(B)
        end
    catch _:_ -> null
    end.

File: backend/lib/energy_gateway/fred_feed_server.ex
defmodule EnergyGateway.FredFeedServer do
  use GenServer

  require Logger

  @name __MODULE__
  @base_url "https://api.stlouisfed.org/fred/series/observations"
  @endpoint :fred

  @series_ids [
    "DTWEXBGS",
    "CPIAUCSL",
    "REAINTRATREARAT10Y",
    "INDPRO",
    "DCOILWTICO"
  ]

  def start_link(_) do
    GenServer.start_link(__MODULE__, nil, name: @name)
  end

  def force_poll do
    GenServer.cast(@name, :poll)
  end

  @impl true
  def init(_) do
    api_key = System.get_env("FRED_API_KEY", "")
    Process.send_after(self(), :poll, 4_000)
    Logger.info("FRED feed started (key present: #{api_key != ""})")
    {:ok, %{interval: 14_400_000, api_key: api_key}}
  end

  @impl true
  def handle_cast(:poll, state) do
    do_poll(state)
    {:noreply, state}
  end

  @impl true
  def handle_info(:poll, %{interval: interval} = state) do
    do_poll(state)
    Process.send_after(self(), :poll, interval)
    {:noreply, state}
  end

  def handle_info(:force_poll, state) do
    do_poll(state)
    {:noreply, state}
  end

  def handle_info(_msg, state), do: {:noreply, state}

  defp do_poll(%{api_key: ""}) do
    Logger.info("FRED API key absent; skipping")
    :ok
  end

  defp do_poll(%{api_key: api_key}) do
    Enum.each(@series_ids, fn series ->
      case EnergyGateway.RateLimiter.check_and_consume(@endpoint) do
        :ok -> fetch_series(series, api_key)
        {:error, :rate_limited} -> Logger.warning("FRED rate limited for #{series}")
      end
    end)
  end

  defp fetch_series(series_id, api_key) do
    url =
      "#{@base_url}?series_id=#{series_id}&api_key=#{api_key}&file_type=json&sort_order=desc&limit=1"

    case http_client().get(url, [], "", [:with_body]) do
      {:ok, 200, _headers, body} -> parse_and_publish(series_id, body)
      {:ok, status, _headers, _body} -> Logger.warning("FRED #{series_id} returned HTTP #{status}")
      {:error, reason} -> Logger.error("FRED #{series_id} error: #{inspect(reason)}")
    end
  end

  defp parse_and_publish(series_id, body) do
    with {:ok, decoded} <- Jason.decode(body),
         %{"observations" => [obs | _]} <- decoded do
      value = Map.get(obs, "value", "")
      num_value = parse_value(value)
      date = Map.get(obs, "date", "")
      series_bin = series_id
      payload = %{"series" => series_bin, "value" => num_value, "date" => date}
      EnergyGateway.EtsCache.put({:macro, series_bin}, payload)
      EnergyGateway.DataRouter.publish(%{
        "type" => "macro",
        "source" => "fred",
        "symbol" => series_bin,
        "timestamp" => System.system_time(:millisecond),
        "payload" => payload
      })
    else
      _ -> Logger.error("FRED parse error for #{series_id}")
    end
  end

  defp parse_value("."), do: nil
  defp parse_value(value) when is_binary(value) do
    case Float.parse(value) do
      {float, _} -> float
      :error -> nil
    end
  end

  defp http_client do
    Application.get_env(:energy_gateway, :http_client, EnergyGateway.HttpClient)
  end
end
